tests_e2e.py

```diff
@@ -0,0 +1,366 @@
+        self.assertEqual (response ['code'], 403)
+
+    def test_ChangePassword (self):
+        user = AuthHelper.getAnyLoggedUser ()
+        response = request ('post', 'auth/change_password', {'cookie': user ['cookie']}, {'old_password': user ['password'], 'new_password': user ['password'] + 'foo'})
+        self.assertEqual (response ['code'], 200)
+
+        # Attempt login with old password
+        response = request ('post', 'auth/login', {}, {'username': user ['username'], 'password': user ['password']})
+        self.assertEqual (response ['code'], 403)
+
+        # Attempt login with new password
+        response = request ('post', 'auth/login', {}, {'username': user ['username'], 'password': user ['password'] + 'foo'})
+        self.assertEqual (response ['code'], 200)
+
+        # Update password on user
+        USERS [user ['username']] ['password'] = user ['password'] + 'foo'
+
+    # TODO MISSING TESTS
+    #def test_Profile (self):
+    #def test_RecoverPassword (self):
+
+# *** TESTS ***
+
+class TestProgram (unittest.TestCase):
+    def test_GetPrograms (self):
+        user = AuthHelper.getAnyLoggedUser ()
+
+        # Get programs but without sending a cookie
+        response = request ('get', 'programs_list', {}, '')
+        # Response should send a redirect to the login page
+        self.assertEqual (response ['code'], 403)
+
+        # Get programs sending a cookie
+        response = request ('get', 'programs_list', {'cookie': user ['cookie']}, '')
+        # Response should be an object of the shape `{programs: [...]}`.
+        self.assertEqual (response ['code'], 200)
+        self.assertIsInstance (response ['body'], dict)
+        self.assertIsInstance (response ['body'] ['programs'], list)
+
+    def test_InvalidCreateProgram (self):
+        user = AuthHelper.getAnyLoggedUser ()
+
+        # Send malformed payloads
+        invalid_payloads = [
+            '',
+            [],
+            {},
+            {'code': 1},
+            {'code': ['1']},
+            {'code': 'hello world'},
+            {'code': 'hello world', 'name': 1},
+            {'code': 'hello world', 'name': 'program 1'},
+            {'code': 'hello world', 'name': 'program 1', 'level': '1'},
+            {'code': 'hello world', 'name': 'program 1', 'level': 1, 'adventure_name': 1},
+        ]
+
+        for invalid_payload in invalid_payloads:
+            response = request ('post', 'programs', {'cookie': user ['cookie']}, invalid_payload)
+            self.assertEqual (response ['code'], 400)
+
+    def test_CreateProgram (self):
+        # We create a new user to ensure that the user has no programs
+        user = AuthHelper.assertUserIsLogged (AuthHelper.makeUsername ())
+
+        program = {'code': 'hello world', 'name': 'program 1', 'level': 1}
+        response = request ('post', 'programs', {'cookie': user ['cookie']}, program)
```

Comment:
We now have to repeat the cookie in every single call to `request('post')`.

If instead we put the state of the current cookies on the current object (on `self`), then we don't need to remember to do this all the time. We could be writing something like this:

```py
   def test_CreateProgram (self):

        # We can hide makeUsername() behind a descriptive name
        # This writes the user and the user's cookies to 'self.user' and 'self.cookie',
        # or some other member variable
        self.given_fresh_user_is_logged_in()

        program = {'code': 'hello world', 'name': 'program 1', 'level': 1}

        # Because `post_data` lives on `self`, it can transparently access self.cookie
        # while doing the POST request. User doesn't need to think about this anymore,
        # cannot get it wrong!
        #
        # Also proposing that `post_data` ALWAYS expects a call to return 200, that saves us
        # an assertEquals check.
        response = self.post_json('/programs', program)

        # If we want to have a response code besides 200, we can have an optional keyword
        # argument for that:
        response = self.post_json('/programs', program, expect_http_code=403)

        # Like 'self.post_json', automaticaly includes the cookie, checks the status code,
        # and reads the body as JSON.
        saved_programs = self.get_json('/programs_list')['programs']

        # Compare two lists of dictionaries
        self.assertEquals(saved_programs, [program])
```